test/convolve_2d_test.cc

```diff
@@ -0,0 +1,534 @@
+                                                     conv_buf_tst_,
+                                                     MAX_SB_SIZE,
+                                                     1,
+                                                     bd);
+                    } else {
+                        conv_params_ref = get_conv_params_no_round(
+                            0, do_average, 0, nullptr, 0, 0, bd);
+                        conv_params_tst = get_conv_params_no_round(
+                            0, do_average, 0, nullptr, 0, 0, bd);
+                    }
+                    // Test special case where dist_wtd_comp_avg is not used
+                    conv_params_ref.use_jnt_comp_avg = 0;
+                    conv_params_tst.use_jnt_comp_avg = 0;
+
+                    const int subx_range = has_subx ? 16 : 1;
+                    const int suby_range = has_suby ? 16 : 1;
+                    for (subx = 0; subx < subx_range; ++subx) {
+                        for (suby = 0; suby < suby_range; ++suby) {
+                            const int offset_r = 3;
+                            const int offset_c = 3;
+                            RunConvolve(offset_r,
+                                        offset_c,
+                                        w,
+                                        MAX_SB_SIZE,
+                                        blk_w,
+                                        blk_h,
+                                        &filter_params_x,
+                                        &filter_params_y,
+                                        subx,
+                                        suby,
+                                        &conv_params_ref,
+                                        &conv_params_tst,
+                                        bd);
+
+                            if (memcmp(
+                                    output_ref_,
+                                    output_tst_,
+                                    MAX_SB_SQUARE * sizeof(output_ref_[0]))) {
+                                for (int i = 0; i < blk_h; ++i) {
+                                    for (int j = 0; j < blk_w; ++j) {
+                                        int idx = i * MAX_SB_SIZE + j;
+                                        ASSERT_EQ(output_ref_[idx],
+                                                  output_tst_[idx])
+                                            << blk_w << "x" << blk_h
+                                            << " Pixel mismatch at index "
+                                            << idx << " = (" << j << ", " << i
+                                            << "), sub pixel offset = (" << suby
+                                            << ", " << subx << ")";
+                                    }
+                                }
+                            }
+                            if (is_jnt) {
+                                for (int i = 0; i < blk_h; ++i) {
```

Comment:
there are similar comparison, is it better write a function to reuse it to make the main function flat and smaller?

Reply:
Good suggestion, modified accordingly. Please review it again.